### Configuración inicial

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, to_date
from delta import *
from pyspark.sql import functions as F
from pyspark.sql.types import DecimalType, TimestampType ,DateType
from delta import configure_spark_with_delta_pip

builder = SparkSession.builder \
    .appName("Lab_SECOP_Silver_Rescate") \
    .master("local[1]") \
    .config("spark.jars.packages", "io.delta:delta-spark_2.12:3.0.0") \
    .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
    .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog") \
    .config("spark.executor.memory", "1g") \
    .config("spark.driver.memory", "1g") \
    .config("spark.memory.offHeap.enabled", "true") \
    .config("spark.memory.offHeap.size", "512m") \
    .config("spark.sql.shuffle.partitions", "1") \
    .config("spark.sql.execution.arrow.pyspark.enabled", "false") # Desactivamos Arrow para evitar picos de RAM

spark = configure_spark_with_delta_pip(builder).getOrCreate()

:: loading settings :: url = jar:file:/opt/spark/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /root/.ivy2/cache
The jars for the packages stored in: /root/.ivy2/jars
io.delta#delta-spark_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-d862a581-e12c-4d5d-8275-84b394a942fd;1.0
	confs: [default]
	found io.delta#delta-spark_2.12;3.0.0 in central
	found io.delta#delta-storage;3.0.0 in central
	found org.antlr#antlr4-runtime;4.9.3 in central
:: resolution report :: resolve 205ms :: artifacts dl 7ms
	:: modules in use:
	io.delta#delta-spark_2.12;3.0.0 from central in [default]
	io.delta#delta-storage;3.0.0 from central in [default]
	org.antlr#antlr4-runtime;4.9.3 from central in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   3   |   0   |   0   |   

### Leer datos de Bronce

In [2]:
spark.catalog.clearCache()

bronze_path = "/app/data/lakehouse/bronze/secop"
df_bronze= spark.read.format("delta").load(bronze_path)

In [3]:
df_bronze.limit(5).toPandas()

26/01/31 01:32:41 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.
26/01/31 01:32:46 WARN GarbageCollectionMetrics: To enable non-built-in garbage collector(s) List(G1 Concurrent GC), users should configure it(them) to spark.eventLog.gcMetrics.youngGenerationGarbageCollectors or spark.eventLog.gcMetrics.oldGenerationGarbageCollectors
                                                                                

anno_bpin      c_digo_bpin         ciudad codigo_de_categoria_principal  \
0      2025  202500000012580        Popayán                   V1.80111600   
1      2025    2024130010059      Cartagena                   V1.80111600   
2      No D      No Definido   Barranquilla                   V1.80111600   
3      2025    2024003050076       Medellín                   V1.80111600   
4      No D      No Definido  Villavicencio                   V1.80111600   

  codigo_entidad codigo_proveedor condiciones_de_entrega departamento  \
0      703038141        723529319             A convenir        Cauca   
1      703301051        732293089            No Definido      Bolívar   
2      702442096        715303566            No Definido    Atlántico   
3      703208645        732329289            No Definido    Antioquia   
4      700817075        709397558             A convenir         Meta   

                             descripcion_del_proceso  \
0  PRESTAR SERVICIOS PROFESIONALES COMO ABOGADO E...   
1  PRESTACION DE SERVICIOS\nPROFESIONALES COMO AB...   
2  AUNAR ESFUERZOS TÉCNICOS; ADMINISTRATIVOS Y FI...   
3  Prestación de servicios de apoyo como tecnólog...   
4  PRESTACION DE SERVICIOS PROFESIONALES ESPECIAL...   

  descripcion_documentos_tipo  ... valor_amortizado valor_de_pago_adelantado  \
0                 No definido  ...                0                        0   
1                 No definido  ...                0                        0   
2                 No definido  ...                0                        0   
3                 No definido  ...                0                        0   
4                 No definido  ...                0                        0   

  valor_del_contrato valor_facturado valor_pagado valor_pendiente_de  \
0           24000000               0            0                  0   
1           16000000               0            0                  0   
2                  0               0            0                  0   
3           28719600        17550867     17550867                  0   
4           36120000        24780000     24780000                  0   

  valor_pendiente_de_ejecucion valor_pendiente_de_pago  \
0                     24000000                24000000   
1                     16000000                16000000   
2                            0                       0   
3                     11168733                11168733   
4                     11340000                11340000   

             _ingestion_time             _source_file  
0 2026-01-31 01:29:15.080936  API_Socrata_Bogota_2025  
1 2026-01-31 01:29:15.080936  API_Socrata_Bogota_2025  
2 2026-01-31 01:29:15.080936  API_Socrata_Bogota_2025  
3 2026-01-31 01:29:15.080936  API_Socrata_Bogota_2025  
4 2026-01-31 01:29:15.080936  API_Socrata_Bogota_2025  

[5 rows x 89 columns]

### Transformación del tipo de dato

In [4]:
# grupos de columnas 

# Números (Money/Decimal)
cols_decimal = [
    "valor_del_contrato", "valor_de_pago_adelantado", "valor_facturado", 
    "valor_pendiente_de_pago", "valor_pagado", "valor_amortizado", 
    "valor_pendiente_de", "valor_pendiente_de_ejecucion", "saldo_cdp", 
    "saldo_vigencia", "presupuesto_general_de_la_nacion_pgn", 
    "sistema_general_de_participaciones", "sistema_general_de_regal_as",
    "recursos_propios_alcald_as_gobernaciones_y_resguardos_ind_genas_",
    "recursos_de_credito", "recursos_propios"
]

#  Fechas (Timestamp)
cols_timestamp = [
    "fecha_de_firma", "fecha_de_inicio_del_contrato", "fecha_de_fin_del_contrato",
    "ultima_actualizacion", "fecha_inicio_liquidacion", "fecha_fin_liquidacion",
    "fecha_de_notificaci_n_de_prorrogaci_n"
]

# números enteros o IDs
cols_int = ["nit_entidad", "codigo_entidad", "dias_adicionados"]

# 3. Aplicar las transformaciones
df_silver_typed = df_bronze

for c in df_bronze.columns:
    if c in cols_decimal:
        df_silver_typed = df_silver_typed.withColumn(c, F.col(c).cast(DecimalType(18, 2)))
    elif c in cols_timestamp:
        df_silver_typed = df_silver_typed.withColumn(c, F.col(c).cast("timestamp"))
    elif c in cols_int:
        df_silver_typed = df_silver_typed.withColumn(c, F.col(c).cast("long"))
    # El resto se quedan como String automáticamente

print("Tipado completo. Revisando esquema...")
df_silver_typed.printSchema()

Tipado completo. Revisando esquema...
root
 |-- anno_bpin: string (nullable = true)
 |-- c_digo_bpin: string (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- codigo_de_categoria_principal: string (nullable = true)
 |-- codigo_entidad: long (nullable = true)
 |-- codigo_proveedor: string (nullable = true)
 |-- condiciones_de_entrega: string (nullable = true)
 |-- departamento: string (nullable = true)
 |-- descripcion_del_proceso: string (nullable = true)
 |-- descripcion_documentos_tipo: string (nullable = true)
 |-- destino_gasto: string (nullable = true)
 |-- dias_adicionados: long (nullable = true)
 |-- documento_proveedor: string (nullable = true)
 |-- documentos_tipo: string (nullable = true)
 |-- domicilio_representante_legal: string (nullable = true)
 |-- duraci_n_del_contrato: string (nullable = true)
 |-- el_contrato_puede_ser_prorrogado: string (nullable = true)
 |-- entidad_centralizada: string (nullable = true)
 |-- es_grupo: string (nullable = true)
 |-- es_pym

### Validación de Datos

In [5]:
# Calcular la duración del contrato mayor a 0
df_with_duration = df_silver_typed.withColumn(
    "duracion_calculada", 
    F.datediff(F.col("fecha_de_fin_del_contrato"), F.col("fecha_de_inicio_del_contrato"))
)

# límite cronológico (Año 1900 adelante)
fecha_limite = "1900-01-01"

# Identificar automáticamente todas las columnas de tipo fecha/timestamp
cols_temporales = [f.name for f in df_with_duration.schema.fields 
                   if isinstance(f.dataType, (DateType, TimestampType))]

# Reglas: 
# - Precio > 0
# - Fecha de Firma no nula
# - Duración > 0
# - Orden diferente a "No Definido"


# Construir la condición dinámica para TODAS las fechas
# Genera: (F.col(fecha1) >= '1900-01-01') & (F.col(fecha2) >= '1900-01-01') 


condicion_fechas_ok = F.lit(True) # Punto de partida neutro para el operador &
for col_name in cols_temporales:
    condicion_fechas_ok &= (F.col(col_name) >= fecha_limite)

# Condición de Validación (Quality Gate)
condicion_valida = (
    (F.col("valor_del_contrato") > 0) & 
    (F.col("fecha_de_firma").isNotNull()) &
    (F.col("fecha_de_firma") >= fecha_limite) &
    (F.col("fecha_de_inicio_del_contrato") >= fecha_limite) &
    (F.col("fecha_de_fin_del_contrato") >= fecha_limite) &
    (F.col("duracion_calculada") > 0) &
    (F.col("orden").isNotNull()) & 
    (F.upper(F.col("orden")) != "NO DEFINIDO")
)

# BIFURCACIÓN (Split)
df_silver = df_with_duration.filter(condicion_valida)

# Registros Inválidos (Quarantine)
df_quarantine = df_with_duration.filter(~condicion_valida)


#  Asignación de Motivos de Rechazo Detallados
df_quarantine = df_quarantine.withColumn(
    "motivo_rechazo",
    F.when(F.col("valor_del_contrato") <= 0, "Precio base menor o igual a 0")
     .when(F.col("fecha_de_firma").isNull(), "Fecha de firma nula")
     .when(F.col("fecha_de_firma") < fecha_limite, "Fecha de firma anterior a 1900")
     .when(F.col("fecha_de_inicio_del_contrato") < fecha_limite, "Fecha inicio anterior a 1900")
     .when(F.col("fecha_de_fin_del_contrato") < fecha_limite, "Fecha fin anterior a 1900")
     .when(F.col("duracion_calculada") <= 0, "Duración de contrato inválida (<= 0 días)")
     .when(F.col("duracion_calculada").isNull(), "Fechas inconsistentes para calcular duración")
     .when(F.upper(F.col("orden")) == "NO DEFINIDO", "Columna 'Orden' marcada como No Definido")
     .when(~condicion_fechas_ok, "Alguna fecha es anterior a 1900 o inválida")
     .otherwise("Error de integridad o campos nulos")
)

# numero de registros
print(f"Registros listos para Silver: {df_silver.count()}")
print(f"Registros para Cuarentena: {df_quarantine.count()}")

Registros listos para Silver: 438143


[Stage 17:=============================>                            (1 + 1) / 2]

Registros para Cuarentena: 6412


### Guardar Datos en Silver y Quarantine

In [6]:
# configuraciones de seguridad para fechas antiguas
spark.conf.set("spark.sql.parquet.int96RebaseModeInWrite", "CORRECTED")
spark.conf.set("spark.sql.parquet.datetimeRebaseModeInWrite", "CORRECTED")

# Rutas
silver_path = "/app/data/lakehouse/silver/secop"
quarantine_path = "/app/data/lakehouse/quarantine/secop_errors"

# Guardado de SILVER 
print(" Iniciando guardado de Silver...")
(df_silver
 .repartition(4) 
 .write
 .format("delta")
 .mode("overwrite")
 .option("maxRecordsPerFile", 50000) # Evita archivos gigantes que consumen mucha RAM
 .save(silver_path))

# Guardado de CUARENTENA
print(" Iniciando guardado de Cuarentena...")
(df_quarantine
 .repartition(2) 
 .write
 .format("delta")
 .mode("overwrite")
 .save(quarantine_path))

print(f" Ingesta Silver exitosa")
print(f" Registros en Cuarentena guardados")

 Iniciando guardado de Silver...


 Iniciando guardado de Cuarentena...


[Stage 40:======================================================> (49 + 1) / 50]

 Ingesta Silver exitosa
 Registros en Cuarentena guardados


### Descriptivas

In [7]:
def resumen_robusto_numericas(df, columnas):
    resultados = []

    for col in columnas:
        # Calculamos las estadísticas
        stats = df.select(
            F.count(F.col(col)).alias("n"),
            F.mean(F.when(F.col(col).isNull(), 1).otherwise(0)).alias("missing_prop"),
            F.expr(f"percentile_approx({col}, 0.5)").alias("mediana"),
            F.expr(f"percentile_approx({col}, 0.1)").alias("p10"),
            F.expr(f"percentile_approx({col}, 0.9)").alias("p90"),
            (
                F.expr(f"percentile_approx({col}, 0.75)") -
                F.expr(f"percentile_approx({col}, 0.25)")
            ).alias("iqr"),
            F.mean(F.when(F.col(col) == 0, 1).otherwise(0)).alias("ceros_prop")
        ).collect()[0]

        # Convertimos TODO a float para evitar el error de mezcla de tipos
        
        resultados.append((
            col,
            int(stats["n"]),
            round(float(stats["missing_prop"]) * 100, 2),
            float(stats["mediana"]) if stats["mediana"] is not None else 0.0,
            float(stats["p10"]) if stats["p10"] is not None else 0.0,
            float(stats["p90"]) if stats["p90"] is not None else 0.0,
            float(stats["iqr"]) if stats["iqr"] is not None else 0.0,
            round(float(stats["ceros_prop"]) * 100, 2)
        ))

    return spark.createDataFrame(
        resultados,
        [
            "variable", "n", "missing_%", "mediana", "p10", "p90", "iqr", "ceros_%"
        ]
    )

vars_numericas = [
    "valor_del_contrato", 
    "valor_pagado", 
    "valor_facturado", 
    "dias_adicionados" ]

resumen_num = resumen_robusto_numericas(df_silver_typed, vars_numericas)
resumen_num.show()

[Stage 64:>                                                         (0 + 1) / 1]

+------------------+------+---------+-----------+---------+-----------+-----------+-------+
|          variable|     n|missing_%|    mediana|      p10|        p90|        iqr|ceros_%|
+------------------+------+---------+-----------+---------+-----------+-----------+-------+
|valor_del_contrato|456617|      0.0|1.4230848E7|4000000.0|    1.174E8|     2.45E7|   0.85|
|      valor_pagado|456617|      0.0|        0.0|      0.0|1.6389504E7|  5974000.0|  66.74|
|   valor_facturado|456617|      0.0|        0.0|      0.0|2.4550833E7|1.1192143E7|  51.36|
|  dias_adicionados|456617|      0.0|        0.0|      0.0|       16.0|        0.0|  86.55|
+------------------+------+---------+-----------+---------+-----------+-----------+-------+



In [8]:
# RESUMEN CATEGORICAS

def resumen_categoricas(df, columnas, top=5):
    resultados = {}
    n_total = df.count()

    for col in columnas:
        tabla = (
            df.groupBy(col)
            .count()
            .withColumn("porcentaje", F.round(F.col("count") / n_total * 100, 2))
            .orderBy(F.desc("count"))
            .limit(top)
        )
        resultados[col] = tabla

    return resultados

# variables de interés
vars_categoricas = [
    "tipo_de_contrato", 
    "modalidad_de_contratacion", 
    "estado_contrato",
    "sector",
    "orden"
]


print("Generando descriptivas categóricas...")
resumen_cat = resumen_categoricas(df_silver_typed, vars_categoricas)

# Imprimir los resultados
for columna, tabla_frecuencia in resumen_cat.items():
    print(f"\n Top 5 para la columna: {columna.upper()}")
    tabla_frecuencia.show(truncate=False)

Generando descriptivas categóricas...



 Top 5 para la columna: TIPO_DE_CONTRATO


+-----------------------+------+----------+
|tipo_de_contrato       |count |porcentaje|
+-----------------------+------+----------+
|Prestación de servicios|359310|78.69     |
|Otro                   |26529 |5.81      |
|Decreto 092 de 2017    |24869 |5.45      |
|Compraventa            |14466 |3.17      |
|Suministros            |14406 |3.15      |
+-----------------------+------+----------+


 Top 5 para la columna: MODALIDAD_DE_CONTRATACION


+------------------------------------+------+----------+
|modalidad_de_contratacion           |count |porcentaje|
+------------------------------------+------+----------+
|Contratación directa                |313201|68.59     |
|Contratación régimen especial       |76965 |16.86     |
|Mínima cuantía                      |32876 |7.2       |
|Selección Abreviada de Menor Cuantía|7677  |1.68      |
|Selección abreviada subasta inversa |7623  |1.67      |
+------------------------------------+------+----------+


 Top 5 para la columna: ESTADO_CONTRATO


+---------------+------+----------+
|estado_contrato|count |porcentaje|
+---------------+------+----------+
|En ejecución   |308634|67.59     |
|Modificado     |97220 |21.29     |
|terminado      |23144 |5.07      |
|Aprobado       |12849 |2.81      |
|Cerrado        |10169 |2.23      |
+---------------+------+----------+


 Top 5 para la columna: SECTOR


+-------------------------+------+----------+
|sector                   |count |porcentaje|
+-------------------------+------+----------+
|Servicio Público         |131838|28.87     |
|No aplica/No pertenece   |81027 |17.75     |
|Salud y Protección Social|76591 |16.77     |
|Educación Nacional       |23163 |5.07      |
|deportes                 |18433 |4.04      |
+-------------------------+------+----------+


 Top 5 para la columna: ORDEN


[Stage 92:=============================>                            (1 + 1) / 2]

+--------------------+------+----------+
|orden               |count |porcentaje|
+--------------------+------+----------+
|Territorial         |330296|72.34     |
|Nacional            |121261|26.56     |
|Corporación Autónoma|5059  |1.11      |
|No Definido         |1     |0.0       |
+--------------------+------+----------+



In [9]:
# ANÁLISIS TEMPORAL (Fechas y Duración)

print("\n Generando análisis temporal...")

# Calculamos la duración en días (Fecha Fin - Fecha Inicio)

df_tiempos = df_silver_typed.withColumn(
    "duracion_contrato_dias",
    F.datediff(F.col("fecha_de_fin_del_contrato"), F.col("fecha_de_inicio_del_contrato"))
)

# rango total de los datos
rango_temporal = df_tiempos.select(
    F.min("fecha_de_firma").alias("Primer_Contrato_Firmado"),
    F.max("fecha_de_firma").alias("Ultimo_Contrato_Firmado"),
    F.min("fecha_de_inicio_del_contrato").alias("Inicio_Ejecucion_Min"),
    F.max("fecha_de_fin_del_contrato").alias("Fin_Ejecucion_Max")
)

# Calculamos la duración promedio y mediana de los contratos
estadisticas_duracion = df_tiempos.select(
    F.round(F.mean("duracion_contrato_dias"), 2).alias("Duracion_Promedio_Dias"),
    F.expr("percentile_approx(duracion_contrato_dias, 0.5)").alias("Duracion_Mediana_Dias"),
    F.min("duracion_contrato_dias").alias("Duracion_Minima"),
    F.max("duracion_contrato_dias").alias("Duracion_Maxima")
)

# Mostrar resultados
print("\n=== RANGO CRONOLÓGICO DE LA DATA ===")
rango_temporal.show()

print("=== ESTADÍSTICAS DE DURACIÓN DE CONTRATOS ===")
estadisticas_duracion.show()


 Generando análisis temporal...

=== RANGO CRONOLÓGICO DE LA DATA ===


+-----------------------+-----------------------+--------------------+-------------------+
|Primer_Contrato_Firmado|Ultimo_Contrato_Firmado|Inicio_Ejecucion_Min|  Fin_Ejecucion_Max|
+-----------------------+-----------------------+--------------------+-------------------+
|    2025-07-01 00:00:00|    2025-12-31 00:00:00| 1899-11-06 00:00:00|5025-12-20 00:00:00|
+-----------------------+-----------------------+--------------------+-------------------+

=== ESTADÍSTICAS DE DURACIÓN DE CONTRATOS ===


[Stage 102:============================>                            (1 + 1) / 2]

+----------------------+---------------------+---------------+---------------+
|Duracion_Promedio_Dias|Duracion_Mediana_Dias|Duracion_Minima|Duracion_Maxima|
+----------------------+---------------------+---------------+---------------+
|                110.34|                   90|        -730477|        1095834|
+----------------------+---------------------+---------------+---------------+



### Análisis de Descriptivos

* La mediana del valor de los contratos es de 14 millones de pesos, con alta dispersión y presencia de contratos muy grandes que alcanzan hasta 117 millones.

* Más del 66% de los contratos no registran pagos, lo que evidencia una brecha entre lo contratado y lo efectivamente desembolsado.

* El 87% de los contratos no presentan adiciones de tiempo; cuando existen, suelen ser menores a 16 días.

* El 79% de los contratos corresponden a prestación de servicios, mostrando un claro predominio sobre otras tipologías.

* El 69% de la contratación se realiza de manera directa, con baja participación de procesos competitivos.

* Dos tercios de los contratos están en ejecución y uno de cada cinco ha sido modificado, mientras que pocos aparecen terminados o cerrados.

* La contratación se concentra en servicios públicos y salud, mientras que educación tiene menor participación.

* El 72% de los contratos son gestionados por entidades territoriales, lo que refleja una marcada descentralización.

* La duración típica de los contratos es de alrededor de 90 días, aunque se identifican registros anómalos que sugieren problemas de calidad en la base de datos.